types/shares.go

```diff
@@ -0,0 +1,148 @@
+// for pushing the shares to the namespaced Merkle tree.
+type NamespacedShare struct {
+	Share
+	ID namespace.ID
+}
+
+func (n NamespacedShare) NamespaceID() namespace.ID {
+	return n.ID
+}
+
+func (n NamespacedShare) Data() []byte {
+	return n.Share
+}
+
+type NamespacedShares []NamespacedShare
+
+// Shares returns the raw shares that can be fed into the erasure coding
+// library (e.g. rsmt2d).
+func (ns NamespacedShares) Shares() []Share {
+	res := make([]Share, len(ns))
+	for i, nsh := range ns {
+		res[i] = nsh.Share
+	}
+	return res
+}
+
+type LenDelimitedMarshaler interface {
+	MarshalDelimited() ([]byte,error)
+}
+
+var _ LenDelimitedMarshaler = Message{}
+var _ LenDelimitedMarshaler = ProtoLenDelimitedMarshaler{}
+var _ LenDelimitedMarshaler = Tx{}
+
+// ProtoLenDelimitedMarshaler decorates any (gogo)proto.Message with a MarshalDelimited
+// method.
+type ProtoLenDelimitedMarshaler struct {
+	proto.Message
+}
+
+func (p ProtoLenDelimitedMarshaler) MarshalDelimited() ([]byte, error) {
+	return protoio.MarshalDelimited(p.Message)
+}
+
+func (tx Tx)  MarshalDelimited() ([]byte, error) {
+	lenBuf := make([]byte, binary.MaxVarintLen64)
+	length := uint64(len(tx))
+	n := binary.PutUvarint(lenBuf, length)
+
+	return append(lenBuf[:n], tx...), nil
+}
+// MarshalDelimited marshals the raw data (excluding the namespace) of this
+// message and prefixes it with the length of that encoding.
+func (m Message) MarshalDelimited() ([]byte, error) {
+	lenBuf := make([]byte, binary.MaxVarintLen64)
+	length := uint64(len(m.Data))
+	n := binary.PutUvarint(lenBuf, length)
+
+	return append(lenBuf[:n], m.Data...), nil
+}
+
+func MakeShares(data []LenDelimitedMarshaler, shareSize int, nidFunc func(elem interface{}) namespace.ID) NamespacedShares {
+	if shareSize <= 0 {
+		panic("invalid shareSize")
+	}
+	shares := make([]NamespacedShare, 0)
+	for _, element := range data {
+		// TODO: implement a helper that also returns the size
+		// to make it possible to (cleanly) implement:
+		// https://github.com/lazyledger/lazyledger-specs/issues/69
+		// For now, we do not squeeze multiple Tx into one share and
+		// hence can ignore prefixing with the starting location of the
+		// first start of a transaction in the share (aka SHARE_RESERVED_BYTES)
+		rawData, err := element.MarshalDelimited()
+		if err != nil {
+			// The (abci) app has to guarantee that it only includes messages
+			// that can be encoded without an error (equivalently tendermint
+			// must not include any Tx, Evidence etc that is unencodable)
+			panic(fmt.Sprintf("can not encode %v", element))
+		}
+		nid := nidFunc(element)
+		if len(rawData) < shareSize {
+			rawShare := rawData
+			paddedShare := zeroPadIfNecessary(rawShare, shareSize)
+			share := NamespacedShare{paddedShare, nid}
+			shares = append(shares, share)
+		} else { // len(rawData) >= shareWithoutNidSize:
+			shares = append(shares, split(rawData, shareSize, nid)...)
+		}
+	}
+	return shares
+}
```

Comment:
@adlerjohn I'll leave away prefixing with the reserved share bytes and initially treat Tx, Evidence etc no different than Messages in the sense that one share will contain one Tx.

Reply:
Meaning, I'm intentionally skipping the `*` in https://github.com/lazyledger/lazyledger-specs/blob/master/specs/figures/share.svg as there is always only one Tx per share in the first version.